src/pstore/pstore.c

```diff
@@ -0,0 +1,735 @@
+#include "strv.h"
+#include "tmpfile-util.h"
+#include "user-util.h"
+#include "util.h"
+
+#define ARRAY_SIZE(ARRAY) (sizeof(ARRAY)/sizeof(ARRAY[0]))
+
+#define PATHSZ 1024
+#define ARG_SOURCEDIR_DEFAULT "/sys/fs/pstore"
+#define ARG_ARCHIVEDIR_DEFAULT "/var/lib/systemd/pstore"
+
+/*
+ * Command line argument handling
+ */
+typedef enum PstoreStorage {
+        PSTORE_STORAGE_NONE,
+        PSTORE_STORAGE_ARCHIVE,
+        PSTORE_STORAGE_JOURNAL,
+        _PSTORE_STORAGE_MAX,
+        _PSTORE_STORAGE_INVALID = -1
+} PstoreStorage;
+
+static const char* const pstore_storage_table[_PSTORE_STORAGE_MAX] = {
+        [PSTORE_STORAGE_NONE] = "none",
+        [PSTORE_STORAGE_ARCHIVE] = "archive",
+        [PSTORE_STORAGE_JOURNAL] = "journal",
+};
+
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP(pstore_storage, PstoreStorage);
+static DEFINE_CONFIG_PARSE_ENUM(config_parse_pstore_storage, pstore_storage, PstoreStorage, "Failed to parse storage setting");
+
+static PstoreStorage arg_storage = PSTORE_STORAGE_ARCHIVE;
+
+static bool arg_allowunlink = true;
+static char *arg_sourcedir = NULL;
+static char *arg_archivedir = NULL;
+STATIC_DESTRUCTOR_REGISTER(arg_sourcedir, freep);
+STATIC_DESTRUCTOR_REGISTER(arg_archivedir, freep);
+
+static int parse_config(void) {
+    int rc;
+    static const ConfigTableItem items[] = {
+            { "Pstore", "AllowUnlink",      config_parse_bool,              0, &arg_allowunlink },
+            { "Pstore", "Storage",          config_parse_pstore_storage,    0, &arg_storage     },
+            { "Pstore", "SourceDir",        config_parse_path,              0, &arg_sourcedir   },
+            { "Pstore", "ArchiveDir",       config_parse_path,              0, &arg_archivedir  },
+            {}
+    };
+
+    rc = config_parse_many_nulstr(PKGSYSCONFDIR "/pstore.conf",
+                                    CONF_PATHS_NULSTR("systemd/pstore.conf.d"),
+                                    "Pstore\0",
+                                    config_item_table_lookup, items,
+                                    CONFIG_PARSE_WARN, NULL);
+    if (NULL == arg_sourcedir)
+    {
+        arg_sourcedir = (char *)malloc(sizeof(ARG_SOURCEDIR_DEFAULT)+1);
+        if (NULL != arg_sourcedir)
+            strcpy(arg_sourcedir, ARG_SOURCEDIR_DEFAULT);
+    }
+    if (NULL == arg_archivedir)
+    {
+        arg_archivedir = (char *)malloc(sizeof(ARG_ARCHIVEDIR_DEFAULT)+1);
+        if (NULL != arg_archivedir)
+            strcpy(arg_archivedir, ARG_ARCHIVEDIR_DEFAULT);
+    }
+    return rc;
+}
+
+/*
+ * File list handling
+ */
+typedef struct dirent_cs
+{
+    struct dirent de;
+    struct stat statbuf;
+    int is_binary;
+    char *contents;
+    struct dirent_cs *next;
+    struct dirent_cs *prev;
+} dirent_ct;
+
+#define END_OF_DIRENT_LIST(LIST, DC) ((DC)->next == (LIST))
+#define START_OF_DIRENT_LIST(LIST, DC) ((DC) == (LIST))
+
+#define FOR_EACH_DIRENT_IN_LIST(LIST, VAR) \
+    for (VAR = LIST; VAR; VAR = (VAR->next == LIST) ? NULL : VAR->next)
```

Comment:
Please use our `LIST` related macros in list.h

Reply:
I'll investigate list handling

Reply:
I've converted to the list.h macros